src/verbs/device_context.rs

```diff
@@ -142,4 +150,89 @@ impl DeviceContext {
             }
         }
     }
+
+    pub fn legacy_query_gid_table(&self) -> Result<Vec<GidEntry>, String> {
```

Comment:
Maybe it is better to not expose the legacy implementation directly to the user; The user is only allowed to invoke `query_gid_table()` and let us handle the transparency among different driver versions.

Reply:
Reasonable, and I think we need to provide some unit tests to make sure the result of legacy and new `query_gid_table` would return the same results.

Reply:
BTW, maybe it is more appropriate to name this method `query_gid_table_fallback` since the original C implementation is called `query_gid_table_fb`